### PR TITLE
Refactor model configuration and parallelization ownership

### PR DESCRIPTION
**Pull request by tianyu-l**
## Human's note

- remove `ModelSpec` and refactor fields into `Model` or `Model.Config`
- remove per-model parallelize.py (including GraphTrainer), and instead use `Model` / `Decoder` / `MultimodalModel` / `FluxModel` to hold the logic, shared with model subclasses
  - introduced `Model._apply_fsdp()` for the above reason
- remove `train_context` as it only calls into `get_spmd_context`
- make `Module.parallelize()` -> `Module._parallelize()` as it's not supposed to be overridden

## Summary

- remove `ModelSpec` and make each model class own its config, checkpoint adapter, optimizer hooks, and high-level parallelization behavior
- consolidate common decoder and multimodal parallelization in `BaseModel`, `Decoder`, and `MultimodalModel`, while keeping `Module._parallelize()` as the low-level primitive
- consolidate the regular GraphTrainer model lifecycle and remove its four per-model `parallelize.py` files
- centralize SPMD context setup and make vocab-parallel log-prob computation take its TP process group explicitly
- define `max_context_length` on decoder configs and set it in concrete model presets

## Why

`ModelSpec` duplicated model identity outside the model class and allowed invalid combinations such as one model with another model checkpoint adapter or optimizer hooks. Per-model parallelization files also repeated behavior that belongs to shared model families. This change makes those relationships structural: selecting a model class determines its configuration and associated behavior, while model subclasses retain focused override points.

Backward compatibility is intentionally not preserved for the removed `ModelSpec` APIs.

## Testing

- `pre-commit run --all-files`
- `pytest tests/unit_tests/cpu -x -q` (1169 passed, 2 skipped)
- affected CPU tests (64 passed)
- focused RL tests (28 passed)
- affected GPU model tests (28 passed, 1 skipped)
- distributed FSDP/sharding tests (9 passed)
- GraphTrainer distributed regression tests (3 passed)
- GraphTrainer test collection (499 tests)
- Flux fake-PG FSDP+compile integration (passed)

The branch is rebased onto current `main`. This is a regular PR so the full upstream CI suite will run.

### CODE REVIEW

File: torchtitan/protocols/model.py [hunk 1]
```diff
@@ -62,34 +72,13 @@ def preprocess_inputs(
...
-        calls this via ``cast(BaseModel, model).preprocess_inputs``,
-        so the declaration lives here for typing. There is no meaningful default;
+        calls this through the ``BaseModel`` interface. There is no meaningful default;
         every model used by the training engine must implement it.
         """
         raise NotImplementedError(
             f"{type(self).__name__} must implement preprocess_inputs()."
         )
 
-    def verify_module_protocol(self) -> None:
```
Comment (tianyu-l) [inline torchtitan/protocols/model.py:73]: capability merged into `_parallelize()` as it's executed recursively for every module already

**Comment by tianyu-l** [comment]
GraphTrainer is the same on main
`
torch.fx.experimental.symbolic_shapes.GuardOnDataDependentSymNode: Could not extract specialized integer from data-dependent expression u41 (unhinted: u41).
`
cc @aditvenk

File: torchtitan/models/deepseek_v4/__init__.py [hunk 2]
```diff
@@ -34,10 +32,6 @@
     make_routed_experts_config,
 )
 from torchtitan.models.common.param_init import depth_scaled_std
-from torchtitan.models.deepseek_v3.parallelize import (
-    parallelize_deepseekv3 as parallelize_deepseek_v4,
```
Comment (jinsooihm) [inline torchtitan/models/deepseek_v4/__init__.py:38]: It seems like removing this got rid of `apply_fsdp_to_mtp_decoder`  used in `parallelize_deepseekv3`. And DSv4 just inherits from `Decoder` not `MTPDecoder`, so the specific fsdp behavior is lost.

File: torchtitan/models/common/multimodal.py [hunk 3]
```diff
@@ -4,17 +4,120 @@
...
+
+from .decoder import Decoder
+
+
+class MultimodalModel(Decoder):
+    """Language model with modality-specific encoders."""
+
+    multimodal_encoder_fqns: tuple[str, ...] = ()
+
+    def parallelize(
```
Comment (wwwjn) [inline torchtitan/models/common/multimodal.py:37]: nit: in human note part, you note that this fuction should be named as `_parallelize()` ?

**Review by wwwjn — APPROVED** [review]
This refactor looks great to me

**Review by aditvenk — APPROVED** [review]

File: torchtitan/models/common/multimodal.py (hunk 3 above)
Comment (tianyu-l) [reply torchtitan/models/common/multimodal.py:37]: `Model.parallelize()` still exist as public entry point
`Module.parallelize()` becomes `Module._parallelize()`

File: torchtitan/models/deepseek_v4/__init__.py (hunk 2 above)
Comment (tianyu-l) [reply torchtitan/models/deepseek_v4/__init__.py:38]: Good catch. DeepSeek V4 was falling back to Decoder._apply_fsdp(), which does not treat its separate mtp_layers as decoder-layer FSDP units. I added a DeepSeekV4Model._apply_fsdp() override that follows MTPDecoder and calls apply_fsdp_to_mtp_decoder(), plus a regression test. Updated in 4c7729ccf.

File: torchtitan/components/loss.py [hunk 4]
```diff
@@ -374,47 +374,45 @@ def compute_logprobs(
...
     When ``return_entropy`` is set, also returns per-token Shannon entropy
     ``H(p) = logsumexp(logits) - sum(softmax(logits) * logits)``, with shape
-    ``[T]``. In a TP SPMD context, ``logits`` holds this rank's local vocab
-    shard. Batch-invariant mode gathers the shards so trainer and vLLM generator
-    perform the same operation sequence. Otherwise, ``global_vocab_size`` must
-    contain the full vocabulary size and statistics are computed directly from
-    the shards.
+    ``[T]``. ``vocab_parallel_group`` explicitly describes the logits layout:
+    a process group means that logits contain a local vocabulary shard, while
+    ``None`` means they contain the full vocabulary. Batch-invariant mode
```
Comment (acisseJZhong) [inline torchtitan/components/loss.py]: not directly related to this PR, but should we remove the mention of vllm?
